crates/runtime-table-partition/src/expression.rs

```diff
@@ -0,0 +1,259 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use arrow_schema::DataType;
+use datafusion::{
+    common::{
+        DFSchema,
+        tree_node::{TreeNode, TreeNodeRecursion},
+    },
+    error::DataFusionError,
+    logical_expr::ExprSchemable,
+    prelude::{Expr, SessionContext},
+};
+use snafu::prelude::*;
+
+#[derive(Debug, Snafu)]
+pub enum ValidationError {
+    #[snafu(display("Failed to determine data type: {source}"))]
+    DataTypeError { source: DataFusionError },
+    #[snafu(display("Expression {expr} does not meet criterion: {message}"))]
```

Comment:
We should list all of the criteria we expect the expression to match here. Its frustrating to be told something is wrong and then fix that one thing and find out something else is wrong.